### PR TITLE
Nodejs env0 cli

### PR DESCRIPTION
**Pull request by omryhay1**
Our Nodejs cli example on how we use env0 API to deploy and destroy environment for each PR.

### CODE REVIEW

**Comment by omryhay1** [comment]
@avnerenv0 - all comments are implemented.
FYI - I couldn't do this `-ev TF_VAR_stage=pr749` as alias can be only one char.

File: node/README.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+# Nodejs client
+
+At [env0](https://env0.com) we believe in eating your own dog food so we are using env0 to deploy env0.
+
+![eat your own dog food]( https://assets.hwvp.com/uploads/articles/eat-your-own-dogfood-saas-vendors-aka-drink-your-own-champagne/_blogFeaturedImage/dogfood.jpg)
+
+This is a Nodejs client that we are using to deploy env0 on each PR.
+
+## Install
```
Comment (avnerenv0) [inline node/README.md]: Add `git clone ...` here

File: node/README.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+``$ npm install``
+
+``$ yarn install``
+
+
+## env0 cli example
+
+  `$ node env0-deploy-cli.js -k apiKey -s apiSecret -a Deploy -o organizationId -p projectId -b blueprintId -e environmentName -r master -v stage=dev`
```
Comment (avnerenv0) [inline node/README.md]:
```suggestion
  `$ node env0-deploy-cli.js -k <apiKey> -s <apiSecret> -a <Deploy> -o <organizationId> -p <projectId> -b <blueprintId> -e <environmentName> -r [master] -v [stage=dev]`
```

File: node/env0-deploy-cli.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+const commandLineUsage = require('command-line-usage');
+const runDeployment = require('./env0-deploy-flow');
+
+const optionDefinitions = [
+  { name: 'apiKey', alias: 'k', type: String, description: 'Your env0 API Key' },
+  { name: 'apiSecret', alias: 's', type: String, description: 'Your env0 API Secret' },
+  { name: 'action', alias: 'a', type: String, description: 'The action you would like to perform - can be Deploy/Destroy' },
+  { name: 'organizationId', alias: 'o', type: String, description: 'Your env0 Organization id' },
+  { name: 'projectId', alias: 'p', type: String, description: 'Your env0 Project id' },
+  { name: 'blueprintId', alias: 'b', type: String, description: 'The Blueprint id you would like to deploy with' },
```
Comment (avnerenv0) [inline node/env0-deploy-cli.js]:
```suggestion
  { name: 'blueprintId', alias: 'b', type: String, description: 'The Blueprint id you would like to deploy' },
```

File: node/env0-deploy-cli.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  { name: 'environmentName', alias: 'e', type: String, description: 'The environment name you would like to create, if it exists it will deploy to that environment' },
+  { name: 'environmentVariables', alias: 'v', type: String, multiple: true, defaultValue: [], description: 'The environment variables to set on the deployed environment - works only on deploy and can be multiple, the format is \"environmentVariableName1=value\"' },
+  { name: 'revision', alias: 'r', type: String, defaultValue: 'master', description: 'You git revision, can be a branch tag or a commit hash. Default value \"master\" ' },
```
Comment (avnerenv0) [inline node/env0-deploy-cli.js]:
```suggestion
  { name: 'revision', alias: 'r', type: String, defaultValue: 'master', description: 'Your git revision, can be a branch tag or a commit hash. Default value \"master\" ' },
```

File: node/env0-deploy-flow.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
+const DeployUtils = require('./env0-deploy-utils');
+
+const deployUtils = new DeployUtils();
+
+const runDeployment = async (options) => {
+    await DeployUtils.init(options);
+    if (options.action === 'Deploy') {
```
Comment (avnerenv0) [inline node/env0-deploy-flow.js]:
```
switch(options.action) {
  case 'Deploy': 
    return createAndDeploy(options);
  case 'Destroy':
    return destory(options);
  default:
    throw new Error(`Action ${options.action} is Invalid, Valid actions are Deploy and Destroy`);
}
```

File: node/env0-deploy-flow.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    console.log('Starting destroying an environment');
+    const environment = await deployUtils.getEnvironment(options.environmentName, options.organizationId);
+    console.log(`got the following environment: ${JSON.stringify(environment)}`);
+
+    if (environment) {
+        await deployUtils.pollEnvironmentStatus(environment.id);
+        await deployUtils.destroyEnvironment(environment);
+        await deployUtils.pollEnvironmentStatus(environment.id);
+    } else {
+        console.log(`No environment found with the name`)
```
Comment (avnerenv0) [inline node/env0-deploy-flow.js]: Maybe throw error here so exit code will be non-zero?

File: node/env0-deploy-flow.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    }
+};
+
+const setConfigurationFromOptions = async (options, environment) => {
+    const environmentVariables = options.environmentVariables;
+    if (environmentVariables && environmentVariables.length > 0) {
+        await environmentVariables.forEach(async(config) => {
+            const configArray = config.split(/=(.+)/);
+            if (configArray.length === 3) {
```
Comment (avnerenv0) [inline node/env0-deploy-flow.js]: This parsing should be done earlier in the flow, ideally by the 'cli' layer, and throw an error if the input is not valid

File: node/env0-deploy-flow.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    await setConfigurationFromOptions(options, environment);
+    await deployUtils.pollEnvironmentStatus(environment.id);
+    await deployUtils.deployEnvironment(environment, options.revision, options.blueprintId);
+    await deployUtils.pollEnvironmentStatus(environment.id);
+};
+
+const destroy = async (options) => {
+    console.log('Starting destroying an environment');
+    const environment = await deployUtils.getEnvironment(options.environmentName, options.organizationId);
+    console.log(`got the following environment: ${JSON.stringify(environment)}`);
```
Comment (avnerenv0) [inline node/env0-deploy-flow.js]: `deployUtils.getEnvironment` already logs this, can be removed

File: node/env0-deploy-flow.js [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    }
+    else {
+        throw new Error(`Action ${options.action} is Invalid, Valid actions are Deploy and Destroy`);
+    }
+};
+
+const createAndDeploy = async (options) => {
+    console.log('Starting deployment');
+    let environment = await deployUtils.getEnvironment(options.environmentName, options.organizationId);
+    console.log(`got the following environment: ${JSON.stringify(environment)}`);
```
Comment (avnerenv0) [inline node/env0-deploy-flow.js]: `deployUtils.getEnvironment` already logs this, can be removed

File: node/env0-deploy-utils.js [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+      name: configurationName,
+      organizationId: environment.organizationId,
+      scope: 2,
+      scopeId: environment.id,
+      type: 0,
+      value: configurationValue
+    };
+
+    console.log(`getting configuration for environmentId: ${environment.id}`);
+    const configurations = await apiClient.callApi('get', `configuration?organizationId=${environment.organizationId}&blueprintId=${blueprintId}&environmentId=${environment.id}`);
```
Comment (avnerenv0) [inline node/env0-deploy-utils.js]:
```suggestion
    const params = { organizationId: environment.organizationId, blueprintId, environmentId: environment.id };
    const configurations = await apiClient.callApi('get', 'configuration', { params });
```

File: node/env0-deploy-flow.js (hunk 7 above)
Comment (omryhay1) [reply node/env0-deploy-flow.js]: Done.

File: node/env0-deploy-flow.js (hunk 9 above)
Comment (omryhay1) [reply node/env0-deploy-flow.js]: Done.

File: node/env0-deploy-flow.js (hunk 8 above)
Comment (omryhay1) [reply node/env0-deploy-flow.js]: Done.

**Comment by omryhay1** [comment]
@avnerenv0 - Implemented all comments.

**Pull request merged**